src/Functions/formatDateTime.cpp

```diff
@@ -538,195 +829,392 @@ class FunctionFormatDateTimeImpl : public IFunction
+                        add_extra_shift(1);
+                        sample += "\t";
                         break;
                     case 'n':
-                        result += '\n';
-                        add_shift(1);
+                        add_extra_shift(1);
+                        sample += "\n";
                         break;
 
                     // Unimplemented
-                    case 'U': [[fallthrough]];
+                    case 'U':
+                        [[fallthrough]];
                     case 'W':
-                        throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                        throw Exception(
+                            ErrorCodes::NOT_IMPLEMENTED,
                             "Wrong pattern '{}', symbol '{}' is not implemented for function {}",
-                            pattern, *pos, getName());
+                            pattern,
+                            *pos,
+                            getName());
 
                     default:
-                        throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                        throw Exception(
+                            ErrorCodes::ILLEGAL_COLUMN,
                             "Wrong pattern '{}', unexpected symbol '{}' for function {}",
-                            pattern, *pos, getName());
+                            pattern,
+                            *pos,
+                            getName());
                 }
 
                 ++pos;
             }
             else
             {
-                result.append(pos, end);
-                add_shift(end + 1 - pos); /// including zero terminator
+                add_extra_shift(end - pos);
+                sample += String(pos, end - pos);
                 break;
             }
         }
 
-        return result;
+        return sample.size();
+    }
+
+    template <typename T>
+    requires (!mysql_pattern)
+    size_t parseJodaPattern(const String & pattern, std::vector<Action<T>> & instructions, String &) const
+    {
+        /// If the argument was DateTime, add instruction for printing. If it was date, just append default literal
+        auto add_instruction = [&](auto && func [[maybe_unused]], const String & default_literal [[maybe_unused]])
+        {
+            if constexpr (std::is_same_v<T, UInt32> || std::is_same_v<T, Int64>)
+                instructions.emplace_back(func);
+            else
+                instructions.emplace_back(std::move(std::bind_front(&Action<T>::template jodaLiteral<String>, default_literal)));
+        };
+
+        size_t reserve_size = 0;
+        const char * cur = pattern.data();
+        const char * end = cur + pattern.size();
+
+        while (cur < end)
+        {
+            const char * start_token_ptr = cur;
+
+            // Literal case
+            if (*start_token_ptr == '\'')
+            {
+                // Case 1: 2 consecutive single quote
+                if (cur + 1 < end && *(cur + 1) == '\'')
```

Comment:
Assuming that the last character of `pattern` is `'` (a quote), then the first condition in l. 1054 fails, we move to the else branch (l. 1061), `numLiteralChars()` in l. 1064 returns 0 and we go into the else branch (l. 1067). I hope my reasoning was right.

The question is should we not end up with an exception in this case?

Reply:
If the last char of `pattern` is `'`(a quote), it must be an ending single quote in the last token. Then the if statement can't be reached because `start_token_ptr == end`.

Reply:
fixed.